refactor(router): RouteData as a type

cexbrayat opened the pull request:
BREAKING CHANGE

The ROUTE_DATA token has been removed and replaced with a type RouteData,
allowing a type injection like we do with RouteParams.

Before:

```
constructor(routeParams: RouteParams, @Inject(ROUTE_DATA) routeData) {
  let id = routeParams.get('id');
  let name = ROUTE_DATA.name;
}
```

After:

```
constructor(routeParams: RouteParams, routeData: RouteData) {
  let id = routeParams.get('id');
  let name = routeData.get('name');
}
```

Fixes #4392

cexbrayat committed:
refactor(router): RouteData as a type

BREAKING CHANGE

The ROUTE_DATA token has been removed and replaced with a type RouteData,
allowing a type injection like we do with RouteParams.

Before:

    constructor(routeParams: RouteParams, @Inject(ROUTE_DATA) routeData) {
      let id = routeParams.get('id');
      let name = ROUTE_DATA.name;
    }

After:

    constructor(routeParams: RouteParams, routeData: RouteData) {
      let id = routeParams.get('id');
      let name = routeData.get('name');
    }

Fixes #4392

btford commented:
Looks good, except two things:

- Can you cache the `RouteData` object on a private field in the instruction so that `instruction.routeData == instruction.routeData`?
- Can you make it a getter?

Fix those, rebase, squash, and this is good to merge :+1:.

Thanks for your help, @cexbrayat!

cexbrayat commented:
@btford Fixed, rebased and squashed.
`_routeData` is now a private field on `ComponentInstruction_` and `routeData()` a getter.
I hope that was what you were expecting. 

I sadly can't figure out how to make the Dart build pass...
Maybe you have an advice? 
It looks like the default parameter value for the `RouteData` constructor is not correct for Dart...

btford commented:
> I sadly can't figure out how to make the Dart build pass

No worries, I can take it from here. Thanks so much!
